File: src/lib/dictionary.ts
import { AVFLAG_NONE } from '../constants/constants.js';
import { bindings } from './binding.js';

import type { AVDictFlag } from '../constants/constants.js';
import type { NativeDictionary, NativeWrapper } from './native-types.js';

/**
 * Key-value dictionary for FFmpeg options and metadata.
 *
 * Stores string key-value pairs used throughout FFmpeg for options, metadata,
 * and configuration. Provides methods for setting, getting, and manipulating
 * dictionary entries. Used extensively for codec options, format options,
 * and metadata handling.
 *
 * Direct mapping to FFmpeg's AVDictionary.
 *
 * @example
 * ```typescript
 * import { Dictionary, FFmpegError } from 'node-av';
 * import { AV_DICT_IGNORE_SUFFIX } from 'node-av/constants';
 *
 * // Create from object
 * const dict = Dictionary.fromObject({
 *   'title': 'My Video',
 *   'artist': 'Me',
 *   'year': '2024'
 * });
 *
 * // Set individual values
 * const dict2 = new Dictionary();
 * dict2.alloc();
 * let ret = dict2.set('preset', 'fast');
 * FFmpegError.throwIfError(ret, 'set');
 *
 * // Parse from string
 * ret = dict2.parseString('key1=value1:key2=value2', '=', ':');
 * FFmpegError.throwIfError(ret, 'parseString');
 *
 * // Get all entries
 * const entries = dict2.getAll();
 * console.log(entries); // { key1: 'value1', key2: 'value2' }
 * ```
 *
 * @see [AVDictionary](https://ffmpeg.org/doxygen/trunk/group__lavu__dict.html) - FFmpeg Doxygen
 * @see {@link CodecContext} For codec options
 * @see {@link FormatContext} For format options
 */
export class Dictionary implements Disposable, NativeWrapper<NativeDictionary> {
  private native: NativeDictionary;

  constructor() {
    this.native = new bindings.Dictionary();
  }

  /**
   * Create dictionary from JavaScript object.
   *
   * Convenience method to create a dictionary from a plain object.
   *
   * @param obj - Object with string key-value pairs
   *
   * @param flags - Flags for setting entries
   *
   * @returns New dictionary with entries from object
   *
   * @example
   * ```typescript
   * const metadata = Dictionary.fromObject({
   *   'title': 'My Song',
   *   'album': 'My Album',
   *   'date': '2024',
   *   'track': '1/10'
   * });
   *
   * // Use for codec options
   * const options = Dictionary.fromObject({
   *   'preset': 'medium',
   *   'crf': 23,
   *   'profile': 'high'
   * });
   * ```
   */
  static fromObject(obj: Record<string, string | number | boolean | bigint | undefined | null>, flags: AVDictFlag = AVFLAG_NONE): Dictionary {
    const dict = new Dictionary();
    dict.alloc();

    for (const [key, value] of Object.entries(obj)) {
      if (value === undefined || value === null) {
        continue;
      }

      dict.set(key, value.toString(), flags);
    }
    return dict;
  }

  /**
   * Create dictionary from native instance.
   *
   * @param native - Native dictionary instance
   *
   * @returns Dictionary wrapper
   *
   * @internal
   */
  static fromNative(native: NativeDictionary): Dictionary {
    const dict = Object.create(Dictionary.prototype) as Dictionary;
    (dict as any).native = native;
    return dict;
  }

  /**
   * Allocate a dictionary.
   *
   * Allocates memory for the dictionary structure.
   * Must be called before using the dictionary.
   *
   * Direct mapping to av_dict_alloc().
   *
   * @throws {Error} If allocation fails (ENOMEM)
   *
   * @example
   * ```typescript
   * const dict = new Dictionary();
   * dict.alloc();
   * // Dictionary is now ready for use
   * ```
   *
   * @see {@link free} To deallocate
   */
  alloc(): void {
    this.native.alloc();
  }

  /**
   * Free the dictionary.
   *
   * Releases all memory associated with the dictionary.
   * The dictionary becomes invalid after calling this.
   *
   * Direct mapping to av_dict_free().
   *
   * @example
   * ```typescript
   * dict.free();
   * // Dictionary is now invalid
   * ```
   *
   * @see {@link alloc} To allocate
   * @see {@link Symbol.dispose} For automatic cleanup
   */
  free(): void {
    this.native.free();
  }

  /**
   * Copy entries to another dictionary.
   *
   * Copies all entries from this dictionary to the destination.
   *
   * Direct mapping to av_dict_copy().
   *
   * @param dst - Destination dictionary
   *
   * @param flags - Copy flags
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - AVERROR_ENOMEM: Memory allocation failure
   *
   * @example
   * ```typescript
   * import { FFmpegError } from 'node-av';
   *
   * const dst = new Dictionary();
   * dst.alloc();
   * const ret = src.copy(dst);
   * FFmpegError.throwIfError(ret, 'copy');
   * ```
   */
  copy(dst: Dictionary, flags: AVDictFlag = AVFLAG_NONE): number {
    return this.native.copy(dst.getNative(), flags);
  }

  /**
   * Set a dictionary entry.
   *
   * Sets or updates a key-value pair in the dictionary.
   *
   * Direct mapping to av_dict_set().
   *
   * @param key - Entry key
   *
   * @param value - Entry value
   *
   * @param flags - Set flags (e.g., AV_DICT_DONT_OVERWRITE)
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - AVERROR_ENOMEM: Memory allocation failure
   *
   * @example
   * ```typescript
   * import { FFmpegError } from 'node-av';
   * import { AV_DICT_DONT_OVERWRITE } from 'node-av/constants';
   *
   * // Set or update entry
   * let ret = dict.set('bitrate', '128k');
   * FFmpegError.throwIfError(ret, 'set');
   *
   * // Set only if not exists
   * ret = dict.set('preset', 'fast', AV_DICT_DONT_OVERWRITE);
   * FFmpegError.throwIfError(ret, 'set');
   * ```
   *
   * @see {@link get} To retrieve values
   */
  set(key: string, value: string, flags: AVDictFlag = AVFLAG_NONE): number {
    return this.native.set(key, value, flags);
  }

  /**
   * Get a dictionary entry.
   *
   * Retrieves the value for a given key.
   *
   * Direct mapping to av_dict_get().
   *
   * @param key - Entry key to look up
   *
   * @param flags - Search flags (e.g., AV_DICT_IGNORE_SUFFIX)
   *
   * @returns Entry value, or null if not found
   *
   * @example
   * ```typescript
   * const value = dict.get('bitrate');
   * if (value) {
   *   console.log(`Bitrate: ${value}`);
   * }
   *
   * // Case-insensitive search
   * import { AV_DICT_MATCH_CASE } from 'node-av/constants';
   * const title = dict.get('Title', AV_DICT_MATCH_CASE);
   * ```
   *
   * @see {@link set} To set values
   * @see {@link getAll} To get all entries
   */
  get(key: string, flags: AVDictFlag = AVFLAG_NONE): string | null {
    return this.native.get(key, flags);
  }

  /**
   * Count dictionary entries.
   *
   * Returns the number of key-value pairs in the dictionary.
   *
   * Direct mapping to av_dict_count().
   *
   * @returns Number of entries
   *
   * @example
   * ```typescript
   * const count = dict.count();
   * console.log(`Dictionary has ${count} entries`);
   * ```
   */
  count(): number {
    return this.native.count();
  }

  /**
   * Get all dictionary entries.
   *
   * Returns all key-value pairs as a JavaScript object.
   *
   * @returns Object with all entries
   *
   * @example
   * ```typescript
   * const entries = dict.getAll();
   * for (const [key, value] of Object.entries(entries)) {
   *   console.log(`${key}: ${value}`);
   * }
   * ```
   *
   * @see {@link get} To get individual entries
   */
  getAll(): Record<string, string> {
    return this.native.getAll();
  }

  /**
   * Parse entries from a string.
   *
   * Parses key-value pairs from a formatted string and adds them
   * to the dictionary.
   *
   * Direct mapping to av_dict_parse_string().
   *
   * @param str - String to parse
   *
   * @param keyValSep - Separator between key and value
   *
   * @param pairsSep - Separator between pairs
   *
   * @param flags - Parse flags
   *
   * @returns 0 on success, negative AVERROR on error:
   *   - AVERROR_EINVAL: Invalid format
   *   - AVERROR_ENOMEM: Memory allocation failure
   *
   * @example
   * ```typescript
   * import { FFmpegError } from 'node-av';
   *
   * // Parse colon-separated pairs
   * let ret = dict.parseString('key1=val1:key2=val2', '=', ':');
   * FFmpegError.throwIfError(ret, 'parseString');
   *
   * // Parse comma-separated pairs
   * ret = dict.parseString('width=1920,height=1080', '=', ',');
   * FFmpegError.throwIfError(ret, 'parseString');
   * ```
   *
   * @see {@link getString} To serialize to string
   */
  parseString(str: string, keyValSep: string, pairsSep: string, flags: AVDictFlag = AVFLAG_NONE): number {
    return this.native.parseString(str, keyValSep, pairsSep, flags);
  }

  /**
   * Convert dictionary to string.
   *
   * Serializes all entries to a formatted string.
   *
   * Direct mapping to av_dict_get_string().
   *
   * @param keyValSep - Separator between key and value
   *
   * @param pairsSep - Separator between pairs
   *
   * @returns Formatted string, or null on error
   *
   * @example
   * ```typescript
   * // Serialize to colon-separated format
   * const str = dict.getString('=', ':');
   * console.log(str); // "key1=val1:key2=val2"
   *
   * // Serialize to comma-separated format
   * const csv = dict.getString('=', ',');
   * console.log(csv); // "key1=val1,key2=val2"
   * ```
   *
   * @see {@link parseString} To parse from string
   */
  getString(keyValSep: string, pairsSep: string): string | null {
    return this.native.getString(keyValSep, pairsSep);
  }

  /**
   * Get the underlying native Dictionary object.
   *
   * @returns The native Dictionary binding object
   *
   * @internal
   */
  getNative(): NativeDictionary {
    return this.native;
  }

  /**
   * Dispose of the dictionary.
   *
   * Implements the Disposable interface for automatic cleanup.
   * Equivalent to calling free().
   *
   * @example
   * ```typescript
   * {
   *   using dict = new Dictionary();
   *   dict.alloc();
   *   dict.set('key', 'value');
   *   // Use dict...
   * } // Automatically freed when leaving scope
   * ```
   */
  [Symbol.dispose](): void {
    this.native[Symbol.dispose]();
  }
}

File: src/bindings/dictionary.cc
#include "dictionary.h"
#include "common.h"

namespace ffmpeg {

thread_local Napi::FunctionReference Dictionary::constructor;

Napi::Object Dictionary::Init(Napi::Env env, Napi::Object exports) {
  Napi::Function func = DefineClass(env, "Dictionary", {
    InstanceMethod<&Dictionary::Alloc>("alloc"),
    InstanceMethod<&Dictionary::Free>("free"),
    InstanceMethod<&Dictionary::Copy>("copy"),
    InstanceMethod<&Dictionary::Set>("set"),
    InstanceMethod<&Dictionary::Get>("get"),
    InstanceMethod<&Dictionary::Count>("count"),
    InstanceMethod<&Dictionary::GetAll>("getAll"),
    InstanceMethod<&Dictionary::ParseString>("parseString"),
    InstanceMethod<&Dictionary::GetString>("getString"),
    InstanceMethod<&Dictionary::Dispose>(Napi::Symbol::WellKnown(env, "dispose")),
  });
  
  constructor = Napi::Persistent(func);
  constructor.SuppressDestruct();
  
  exports.Set("Dictionary", func);
  return exports;
}

Dictionary::Dictionary(const Napi::CallbackInfo& info)
  : Napi::ObjectWrap<Dictionary>(info) {
  // Constructor does nothing - user must explicitly call alloc() or set()
}

Dictionary::~Dictionary() {
  av_dict_free(&dict_);
}

Napi::Value Dictionary::Alloc(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  // FFmpeg doesn't have explicit av_dict_alloc, it's implicit in av_dict_set
  // Free any existing dictionary first
  av_dict_free(&dict_);

  return env.Undefined();
}

Napi::Value Dictionary::Free(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  av_dict_free(&dict_);
  return env.Undefined();
}

Napi::Value Dictionary::Copy(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 2 || !info[0].IsObject() || !info[1].IsNumber()) {
    Napi::TypeError::New(env, "Dictionary and flags required").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  // Get destination dictionary
  Napi::Object dstObj = info[0].As<Napi::Object>();
  Dictionary* dst = UnwrapNativeObject<Dictionary>(env, dstObj, "Dictionary");
  
  if (!dst) {
    Napi::TypeError::New(env, "Invalid dictionary object").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  int flags = info[1].As<Napi::Number>().Int32Value();
  
  // Copy from this dictionary to destination
  // av_dict_copy merges the dictionaries, it doesn't replace
  int ret = av_dict_copy(&dst->dict_, dict_, flags);

  return Napi::Number::New(env, ret);
}

Napi::Value Dictionary::Set(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 3 || !info[0].IsString() || !info[1].IsString() || !info[2].IsNumber()) {
    Napi::TypeError::New(env, "Key (string), value (string), and flags (number) required").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  std::string key = info[0].As<Napi::String>().Utf8Value();
  std::string value = info[1].As<Napi::String>().Utf8Value();
  int flags = info[2].As<Napi::Number>().Int32Value();
  
  // av_dict_set will allocate the dictionary if it's NULL
  int ret = av_dict_set(&dict_, key.c_str(), value.c_str(), flags);

  return Napi::Number::New(env, ret);
}

Napi::Value Dictionary::Get(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 2 || !info[0].IsString() || !info[1].IsNumber()) {
    Napi::TypeError::New(env, "Key (string) and flags (number) required").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  if (!dict_) {
    return env.Null();
  }
  
  std::string key = info[0].As<Napi::String>().Utf8Value();
  int flags = info[1].As<Napi::Number>().Int32Value();
  
  // av_dict_get with NULL as prev to get first match
  AVDictionaryEntry* entry = av_dict_get(dict_, key.c_str(), nullptr, flags);
  
  if (!entry) {
    return env.Null();
  }
  
  return Napi::String::New(env, entry->value);
}

Napi::Value Dictionary::Count(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (!dict_) {
    return Napi::Number::New(env, 0);
  }
  
  int count = av_dict_count(dict_);
  return Napi::Number::New(env, count);
}

Napi::Value Dictionary::GetAll(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  Napi::Object result = Napi::Object::New(env);
  
  if (!dict_) {
    return result;
  }
  
  AVDictionaryEntry* entry = nullptr;
  while ((entry = av_dict_get(dict_, "", entry, AV_DICT_IGNORE_SUFFIX))) {
    result.Set(entry->key, Napi::String::New(env, entry->value));
  }
  
  return result;
}

Napi::Value Dictionary::ParseString(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 4 || !info[0].IsString() || !info[1].IsString() || 
      !info[2].IsString() || !info[3].IsNumber()) {
    Napi::TypeError::New(env, "String, keyValSep (string), pairsSep (string), and flags (number) required")
      .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  std::string str = info[0].As<Napi::String>().Utf8Value();
  std::string keyValSep = info[1].As<Napi::String>().Utf8Value();
  std::string pairsSep = info[2].As<Napi::String>().Utf8Value();
  int flags = info[3].As<Napi::Number>().Int32Value();
  
  // Free old dictionary if exists
  av_dict_free(&dict_);

  int ret = av_dict_parse_string(&dict_, str.c_str(), keyValSep.c_str(), pairsSep.c_str(), flags);

  return Napi::Number::New(env, ret);
}

Napi::Value Dictionary::GetString(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();
  
  if (info.Length() < 2 || !info[0].IsString() || !info[1].IsString()) {
    Napi::TypeError::New(env, "keyValSep (string) and pairsSep (string) required")
      .ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  if (!dict_) {
    return Napi::String::New(env, "");
  }
  
  std::string keyValSep = info[0].As<Napi::String>().Utf8Value();
  std::string pairsSep = info[1].As<Napi::String>().Utf8Value();
  
  if (keyValSep.length() != 1 || pairsSep.length() != 1) {
    Napi::TypeError::New(env, "Separators must be single characters").ThrowAsJavaScriptException();
    return env.Undefined();
  }
  
  char* buffer = nullptr;
  int ret = av_dict_get_string(dict_, &buffer, keyValSep[0], pairsSep[0]);
  
  if (ret < 0 || !buffer) {
    return env.Null();
  }
  
  Napi::String result = Napi::String::New(env, buffer);
  av_free(buffer);
  
  return result;
}

Napi::Value Dictionary::Dispose(const Napi::CallbackInfo& info) {
  return Free(info);
}

} // namespace ffmpeg